### ISSUE TITLE
Using '.where' with 'knex.raw' statements and '.fetchOne' from bookshelf 0.8.1 causes an error.

### ISSUE THREAD

**Issue by wilsontgh** [tgriesser/knex#902]
```
{ [error: select "cm_vehicles".* from "cm_vehicles" where details ->> 'last_fine_check' < $1 or (details ->> 'last_fine_check') is null order by (details ->> 'last_fine_check') ASC limit $2 - bind message supplies 1 parameters, but prepared statement "" requires 2]
  name: 'error',
  severity: 'ERROR',
  code: '08P01',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  file: 'postgres.c',
  line: '1500',
  routine: 'exec_bind_message' }
```
Here are the statements:
```
qb.where(knex.raw("details ->> 'last_fine_check' < ?", oneDayAgo))
			.orWhere(knex.raw("(details ->> 'last_fine_check') is null"))
			.orderBy(knex.raw("(details ->> 'last_fine_check')"), "ASC");

Model.Vehicle.collection().query(query).fetchOne().then(...)
```
This happens if I use knex.js 0.8.6 but does not happen in knex.js 0.7.6.

**Comment by rhys-vdw** [comment]
Hi @wilsontgh, sorry for the late reply.

The problem here is that `.raw` requires its bindings in an array.

This should work:

```js
qb.where(knex.raw("details ->> 'last_fine_check' < ?", [oneDayAgo]))
            .orWhere(knex.raw("(details ->> 'last_fine_check') is null"))
            .orderBy(knex.raw("(details ->> 'last_fine_check')"), "ASC");
```

An argument could certainly be made for raising an error if a non-array value is supplied as the "bindings" argument for `raw`. Either way the documentation should be made more clear.

By the way, you can simplify your code like so:

```js
qb.whereRaw("details ->> 'last_fine_check' < ?", [oneDayAgo])
  .orWhereRaw("(details ->> 'last_fine_check') is null")
  .orderByRaw("(details ->> 'last_fine_check') asc");
```

**Comment by rhys-vdw** [comment]
Request PR for the following changes:

 - Update documentation to clarify that `knex.raw(expression, bindings)` accepts only an array as its `bindings` argument.
 - Raises an error if bindings is supplied other than `null`, `undefined` or an instance of `Array`.

I considered that `.raw` could also accept a single value for `bindings`. However, it is possible to supply an array as one of the bindings, so that would create ambiguity.

**Comment by chuym** [comment]
@rhys-vdw I'm working on a PR for this. I notice [there is a test](https://github.com/tgriesser/knex/blob/5003b41dae3ebbc0e6d439c6eabd6d315a799951/test/unit/query/builder.js#L899-910) that uses a plain string on a raw query. Throwing an error breaks this test so there seems to be some cases where the raw expression works when passing anything other than an array. `raw.js` [handles](https://github.com/tgriesser/knex/blob/5003b41dae3ebbc0e6d439c6eabd6d315a799951/src/raw.js#L46) cases when the bindings are not an array.

I encountered this problem when passing a date object on the query, it works fine when not using `raw` but looking at the `toSQL` method on `raw.js` it seems that when a plain object is passed, it iterates over the values and handles accordingly, but if you pass anything other than a plain object, it tries to do the same and ends up "silently" failing, I say silently failing because the procedure returns `''` when passing objects that are not plain object. I think the right fix would be to check if `bindings` is a plain object, but I might be missing something. What do you think?

**Comment by rhys-vdw** [comment]
Closed by #1051.
